beacon_node/http_api/src/lib.rs

```diff
@@ -1179,6 +1180,7 @@ pub fn serve<T: BeaconChainTypes>(
              accept_header: Option<api_types::Accept>| {
                 async move {
                     let (block, execution_optimistic) = block_id.full_block(&chain).await?;
+                    let finalized = block_id.is_finalized(&chain);
```

Comment:
In the cases where we have a `block_id` we also usually have the `block` already loaded, so it's wasteful to load it again.

To accomplish this I think the method on `BeaconChain` should look something like this:

```rust
pub fn is_finalized_block(&self, block_root: &Hash256, block_slot: Slot) -> Result<bool, Error>;
```

The implementation could check that `block_slot` is less than or equal to the finalized slot _and_ that the canonical root at `block_slot` matches `block_root`. The canonical check is necessary because the database may have some non-canonical blocks and states which conflict with finalization.

```rust
let finalized_slot = chain
    .canonical_head
    .cached_head()
    .finalized_checkpoint()
    .epoch
    .start_slot(E::slots_per_epoch());

let is_canonical = chain
    .block_root_at_slot(block_slot, WhenSlotSkipped::None)?
    .map_or(false, |canonical_root| block_root == canonical_root);

block_slot <= finalized_slot && is_canonical
```

For states something similar is required (check the slot _and_ the canonical status).

If we want to optimize these a little we can also return an optional finality status from a method like `BlockId::root_and_finality_status` returning `Result<(Hash256, Option<bool>), Error>`

Reply:
I'm confused here. My current understanding is:
As the Beacon-API docs say, `block_id`s can be one of: "head", "genesis", "finalized", \<slot\>, \<hex encoded blockRoot with 0x prefix\>. Those variants are represented in the `CoreBlockId` enum that's wrapped into the `BlockId` tuple struct.

So, when someone calls a path that includes a `block_id` we'd either have a slot _or_ a root.

For this reason, I don't understand how can I implement a function that takes both the slot and the root of the block I want to check. If I start with the slot, I'd have to call `chain.block_root_at_slot(slot,...)`, to get a root I can pass to the function. This would beat the purpose of passing both the slot and the root because the root would always be canonical. If I start with the root, I'd have to call `chain.get_blinded_block(root)` to get a slot, which again, would beat the purpose of checking if it's canonical because the slot would always be calculated from the root.

Reply:
Also, in what sense is the block already loaded when having the `block_id`?

Reply:
For the `GET beacon/blocks/{block_id}` API I mean we've already loaded the block on the previous line:

```rust
let (block, execution_optimistic) = block_id.full_block(&chain).await?;
```

So I was imagining checking that it's finalized like so:

```rust
let block_root = block_id.root(&chain)?;
let finalized = chain.is_finalized_block(block_root, block.slot())?;
```

This avoids repeating an expensive database read loading the block from the database a second time.

> If I start with the slot, I'd have to call chain.block_root_at_slot(slot,...), to get a root I can pass to the function.

This observation is correct: when we do lookups by slot we almost always return the canonical block, so checking that it's canonical in this case is redundant (we just need to check if it's prior to the finalized). Structuring the code to allow this would be a nice optimisation, but not 100% necessary.

> If I start with the root, I'd have to call chain.get_blinded_block(root) to get a slot, which again, would beat the purpose of checking if it's canonical because the slot would always be calculated from the root.

This isn't quite right though: there's only 1 canonical block root per slot, but multiple block roots _may_ point to blocks with the same slot. So we do still need to check whether the root is canonical (but _do not_ need to load the block more than once).

Some form of the `block` is loaded by all the APIs which accept a `{block_id}`, so similar logic can be applied for all of them.

I hope that makes sense, let me know if anything is still unclear :blush:

Reply:
That made everything clear, thank you! Pls see my latest commit to check if I missed something :)

Also, should I delete my original `is_finalized()`? I think it still could be useful in case we don't have the `block` loaded.

And one minor thing: I think the name `is_block_finalized` is a bit cleaner than `is_finalized_block`, is there a reason you choose that name specifically? haha

Reply:
Latest commit looks great!

> Also, should I delete my original is_finalized()? I think it still could be useful in case we don't have the block loaded.

I think we may as well delete it, as far as I can tell all the APIs have the block already loaded, even the `/root` API.

> And one minor thing: I think the name is_block_finalized is a bit cleaner than is_finalized_block, is there a reason you choose that name specifically?

I don't mind either way, I chose `is_finalized_block` because it's quite common to use finalized as a prefix adjective like "download a finalized block", etc. But `is_block_finalized` is fine too